staging/src/k8s.io/apiserver/pkg/server/serve.go

```diff
@@ -224,22 +231,222 @@ func GetNamedCertificateMap(certs []NamedTLSCert) (map[string]*tls.Certificate,
+// assume if an error is returned from a chain handler that it will return a nil
+// conn and have already closed that conn. Init() is used to allow low levels tweaks
+// on the TCPConn.
+type tcpConnectionChainHandler interface {
+	Init(*net.TCPConn) error
+}
+
+type tcpConnectionChainListener struct {
+	*net.TCPListener
+	chain []tcpConnectionChainHandler
+}
+
+func (l tcpConnectionChainListener) Accept() (net.Conn, error) {
+	tc, err := l.AcceptTCP()
+	if err != nil {
+		return nil, err
+	}
+	for _, handler := range l.chain {
+		err = handler.Init(tc)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return tc, nil
+}
+
 // tcpKeepAliveListener sets TCP keep-alive timeouts on accepted
 // connections. It's used by ListenAndServe and ListenAndServeTLS so
 // dead TCP connections (e.g. closing laptop mid-download) eventually
 // go away.
 //
 // Copied from Go 1.7.2 net/http/server.go
-type tcpKeepAliveListener struct {
-	*net.TCPListener
+type tcpKeepAliveHandler struct{}
+
+func (ln tcpKeepAliveHandler) Init(conn *net.TCPConn) error {
+	conn.SetKeepAlive(true)
+	conn.SetKeepAlivePeriod(defaultKeepAlivePeriod)
+	return nil
+}
+
+// IPBasedLimitListener accepts at most n simultaneous connections from a
+// given IP address. We also limit the total simultaneous connections.
+// None of this is applied to localhost/loopback connections.
+//
+// Based on Go 1.7.4 net/netutil/listen.go
+// TODO: (cheftako) : IPv6, CIDR Range rather than individual IP Addresses?
+// Considered using atomic for per IP limit value,
+// but problem initing map value safely
+type IPBasedLimit struct {
+	// The maximum concurent connections an IP is allowed to have open
+	MaxPerIP int
+
+	// The total maximum concurrent connections allowed open (Except localhost)
+	Max int
+
+	// Lock used to ensure the Limits and total are updated atomicaly
+	mutex sync.Mutex
+
+	// Map in which we map the serialized ip to the current count
+	Limits map[string]int
+
+	// Map in which we retain the lookup from connection to decrementIP call
+	Decrementers map[string]func()
+
+	// Current total count of the connections in existance except those from localhost
+	total int
+}
+
+func (i *IPBasedLimit) incrementIP(conn, sourceIP string) error {
+	if i.Max <= 0 {
+		return nil
+	}
+	ip := net.ParseIP(sourceIP)
+	if ip == nil {
+		glog.Infof("Received request from client with invalid ip %s", sourceIP)
+		return IPLimitExceededError{ip: sourceIP, reason: "unparseable ip address"}
+	}
+	if ip.IsLoopback() {
+		return nil
+	}
+	i.mutex.Lock()
+	defer i.mutex.Unlock()
+	current := i.Limits[sourceIP]
+	if current >= i.MaxPerIP {
+		return IPLimitExceededError{ip: sourceIP, reason: fmt.Sprintf("exceeded per IP limit of %d", i.MaxPerIP)}
+	}
+	if i.total >= i.Max {
+		return IPLimitExceededError{ip: sourceIP, reason: fmt.Sprintf("exceeded overall limit of %d", i.Max)}
+	}
+	i.total++
+	i.Limits[sourceIP] = current + 1
+	var releaseOnce sync.Once
+	i.Decrementers[conn] = func() {
```

Comment:
I see us locking here to write to Decrementers, but not locking to read in HandleHttpServerStateChange... how does that not cause races?